pkg/devfile/adapters/kubernetes/utils/utils.go

```diff
@@ -94,28 +95,44 @@ func GetContainers(devfileObj devfileParser.DevfileObj) ([]corev1.Container, err
 		// If `mountSources: true` was set, add an empty dir volume to the container to sync the source to
 		// Sync to `Container.SourceMapping` if set
 		if comp.Container.MountSources {
-			var syncFolder, projectsRoot string
+			var syncRootFolder, projectsRoot string
 			if comp.Container.SourceMapping != "" {
-				syncFolder = comp.Container.SourceMapping
+				syncRootFolder = comp.Container.SourceMapping
 			} else if projectsRoot = adaptersCommon.GetComponentEnvVar(adaptersCommon.EnvProjectsRoot, comp.Container.Env); projectsRoot != "" {
-				syncFolder = projectsRoot
+				syncRootFolder = projectsRoot
 			} else {
-				syncFolder = kclient.OdoSourceVolumeMount
+				syncRootFolder = kclient.OdoSourceVolumeMount
 			}
 
 			container.VolumeMounts = append(container.VolumeMounts, corev1.VolumeMount{
 				Name:      kclient.OdoSourceVolume,
-				MountPath: syncFolder,
+				MountPath: syncRootFolder,
 			})
 
 			// only add the env if it is not set by the devfile
 			if projectsRoot == "" {
 				container.Env = append(container.Env,
 					corev1.EnvVar{
 						Name:  adaptersCommon.EnvProjectsRoot,
+						Value: syncRootFolder,
+					})
+			}
+
+			if syncFolder := adaptersCommon.GetComponentEnvVar(adaptersCommon.EnvProjectsSrc, comp.Container.Env); syncFolder == "" {
+				syncFolder, err = sync.GetSyncFolder(syncRootFolder, devfileObj.Data.GetProjects())
+				if err != nil {
+					return nil, err
+				}
+
+				container.Env = append(container.Env,
+					corev1.EnvVar{
+						Name:  adaptersCommon.EnvProjectsSrc,
 						Value: syncFolder,
 					})
+			} else {
+				return nil, fmt.Errorf("env variable %s is reserved and cannot be customized in component %s", adaptersCommon.EnvProjectsSrc, comp.Name)
```

Comment:
Could we error out earlier if `PROJECT_SOURCE` is set in the devfile? Maybe as part of validation?

Reply:
Yeah I thought about this. But I was wondering if for some reason, devfile parse validation changes in the future and it removed the env check; then we'd somehow allow this when creating a container. So put it as double fail safe. I know its redundant but i can remove it, if you feel strongly

Reply:
IMO, we just need the one check for now to avoid having duplicate code (and code that won't get triggered currently), and if anyone overhauls the validation code in the future, it's their responsibility to make sure the check for `PROJECT_SOURCE` still occurs

But I'm totally flexible either way, TBH

Reply:
With devfile parser being extracted out, would we want to validate this at the parser level or push level? 🤔

Reply:
As part of `ValidateDevfileData` I guess? https://github.com/openshift/odo/blob/master/pkg/devfile/validate/validate.go#L13

Reply:
Opened https://github.com/devfile/api/issues/132 for discussion too, FYI. 

Your point about moving out the devfile parser reminded me that we don't really have anything in the spec right now that says `PROJECT_SOURCE` can't be overridden.